src/main/java/org/jooq/lambda/Seq.java

```diff
@@ -231,26 +247,43 @@
      * <p>
      * <code><pre>
      * // (tuple(1, 1), tuple(2, 2))
-     * Seq.of(1, 2, 3).innerJoin(Seq.of(1, 2), t -> Objects.equals(t.v1, t.v2))
+     * Seq.of(1, 2, 3).innerJoin(Seq.of(1, 2), (t, u) -> Objects.equals(t, u))
      * </pre></code>
      */
     default <U> Seq<Tuple2<T, U>> innerJoin(Seq<? extends U> other, BiPredicate<? super T, ? super U> predicate) {
 
         // This algorithm isn't lazy and has substantial complexity for large argument streams!
         List<? extends U> list = other.toList();
 
-        return flatMap(t -> seq(list)
-              .filter(u -> predicate.test(t, u))
-              .map(u -> tuple(t, u)))
+        return flatMap(t -> seq(list).filter(u -> predicate.test(t, u))
+                                     .map(u -> tuple(t, u)))
               .onClose(other::close);
     }
 
     /**
+     * Inner join stream with itself.
+     * <p>
+     * <code><pre>
+     * // (tuple(1, 1), tuple(2, 2))
+     * Seq.of(1, 2).innerSelfJoin((t, u) -> Objects.equals(t, u))
+     * </pre></code>
+     */
+    default Seq<Tuple2<T, T>> innerSelfJoin(BiPredicate<? super T, ? super T> predicate) {
+
+        // This algorithm isn't lazy and has substantial complexity for large argument streams!
+        List<? extends T> list = toList();
+
+        return seq(list).flatMap(t -> seq(list)
+                                     .filter(u -> predicate.test(t, u))
+                                     .map(u -> tuple(t, u)));
+    }
```

Comment:
Hmm, I wonder if we perhaps would be better off if this method delegates to the existing `innerJoin()` method, and we start optimising these joins in a different task...